Source/Dafny/Triggers/QuantifiersCollection.cs

```diff
@@ -299,32 +307,42 @@ internal class QuantifierGroup {
         }
 
         AddTriggersToMessage("Selected triggers:", q.Candidates, msg, indent);
+        if (q.Candidates.Count > 0) {
+          SetRelatedLocationMessage("Trigger position:");
```

Comment:
If there are multiple trigger it doesn't make sense to point to one of them in particular.

Reply:
After reading carefully your comment, I investigated what are exactly these nested tokens, because although they coincide with triggers, they are not covering all triggers since there is only one.
It turns out, the nested token does not describe triggers, it describes only the considered sub-expression after splitting it.

For example, if we had `forall a in A :: (R(a) ==> P(a)) && !(R(a) ==> Q(a))`, then it will first split this expression into three comprehensions `forall a in A :: R(a) ==> P(a))`, `forall a in A :: R(a)` and `forall a in A :: !Q(a)` (and this "not" is reusing the same token as its argument), and then it will look for trigger terms in each sub-expressions. However, triggers _candidates_ are a set of trigger terms, not just terms, so the position this "Related position" gives is actually the positions of the sub-expression that cannot be split further (e.g.  in our case `R(a) ==> P(a)`, `R(a)` and `Q(a)`

Hence, instead of calling them "trigger position", it would be more appropriate to call them "While considering sub-expression at position" if this sub-expression is not the body of the original entire forall expression.

Reply:
Ah, that makes a lot more sense!
Don't we already report that information?  For example this:

```
predicate P(x: int)
predicate Q(x: int)

function f() : int
  requires forall x: int :: P(x) && (Q(x) ==> P(x+1))
```

…gives me the following message:

```
For expression "P(x)":
  Selected triggers: {P(x)}
For expression "Q(x) ==> P(_t#0)":
   Selected triggers: {P(_t#0), Q(x)}
```

Reply:
Annoyingly that message seems to have regressed, since originally it said `P(x)` instead of `P(_t#0)`

Reply:
One thing to consider is that, once the message is built, it is passed to "reporter.Message", and we don't know which implementation it is (console or IDE). In the IDE, and later also for Boogie, we do want this information positionally if it is useful (i.e. here is a header, which means there are more than one splitted sub-expression)
Indeed, the NestedToken is used by boogie to report "Related location" which point to the sub-expression that is being considered, using a single click, so we definitely want to keep the NestedToken with its message.

Concerning the command-line, I think it's acceptable to have more information about the position of sub-expressions, especially since those sub-expressions might not be exactly the same as the original expressions (i.e. the "not" example")

```
For expression "P(x)":
  Selected triggers: {P(x)} in subexpression at document.dfy(5, 29)
For expression "Q(x) ==> P(_t#0)":
   Selected triggers: {P(_t#0), Q(x)} in subexpression at document.dfy(5, 42)
```

What do you think of the above?

I can add your test and verify that it still works (and let me see if I can fix this hideous generated variable, which does make sense since it's not x but x+1)

Reply:
I also hesitate between

* `in sub-expression at`
* `at considered expression position:`
* `inside the expression at`
* `after considering the expression at`

Which one would you think is better?